### PR TITLE
Fix flowchart completion after faulted child activity cancellation

### PR DESCRIPTION
**Pull request by sfmskywalker**
This PR fixes a number of issues:

- [x] When using the ModifyVariable alteration to "fix" the input, the value does not convert into an `ExpandoObject`, but rather a JsonElement; this causes the Fault activity to execute again.
- [x] Rescheduling a workflow activity doesn't work - the internal logic to try and reschedule this activity within its containing flowchart fails to find the flowchart. Instead, rescheduling the previous WriteLine activity does work.
- [x] Canceling the faulted activity does not invoke the container Flowchart activity of this event, which prevents this Flowchart activity from completing itself.

### CODE REVIEW

**Review by MariusVuscanNx — APPROVED** [review]

File: src/modules/Elsa.Workflows.Core/Services/WorkflowRunner.cs [hunk 1]
```diff
@@ -66,15 +67,19 @@ public async Task<RunWorkflowResult<TResult>> RunAsync<TResult>(WorkflowBase<TRe
     }
 
     /// <inheritdoc />
-    public async Task<RunWorkflowResult> RunAsync<T>(RunWorkflowOptions? options = default, CancellationToken cancellationToken = default) where T : IWorkflow
+    public async Task<RunWorkflowResult> RunAsync<[DynamicallyAccessedMembers(DynamicallyAccessedMemberTypes.PublicParameterlessConstructor)] T>(
```
Comment (raymonddenhaan) [inline src/modules/Elsa.Workflows.Core/Services/WorkflowRunner.cs:70]: We need this attribute due to an underlying function, which means if we want to be consistent we need to propogate this all the way up, cluttering a lot of our code with these attributes.
Instead of this attribute we can also put the new() constraint on T

File: src/modules/Elsa.Alterations.Core/Services/WorkflowInstanceFinder.cs [hunk 2]
```diff
@@ -52,10 +52,24 @@ public async Task<IEnumerable<string>> FindAsync(AlterationWorkflowInstanceFilte
...
                 workflowInstanceIds.IntersectWith(matchingWorkflowInstanceIds);
         }
+        
+        // Alterations must apply only to running workflows.
+        workflowInstanceIds = (await FilterRunningWorkflowInstancesAsync(workflowInstanceIds, cancellationToken)).ToHashSet();
 
         return workflowInstanceIds;
     }
 
+    private async Task<IEnumerable<string>> FilterRunningWorkflowInstancesAsync(IEnumerable<string> workflowInstanceIds, CancellationToken cancellationToken)
```
[the remark this answers is not in the document]
Comment (sfmskywalker) [reply src/modules/Elsa.Alterations.Core/Services/WorkflowInstanceFinder.cs:62]: Thank you. I wasn't sure I used the right naming. One the one hand, I was like: "I am filtering _out_ the other states". But then, what remains is the filtered _in_ results. In other words, from what perspective do you look at the term "filtered"? I will rename it then form the perspective of "to filter _out_ the not wanted states.

[the remark this answers is not in the document]
Comment (sfmskywalker) [reply src/modules/Elsa.Alterations.Core/Services/WorkflowInstanceFinder.cs:62]: So I asked ChatGPT about this situation. It actually makes the claim that in programming, we focus on _what to include_. Here's the explanation:

"In programming and data processing, the term "filter" is typically understood to mean the selection of elements that meet a certain condition, while excluding those that do not. The focus is on what is being included rather than what is being excluded. Therefore, if you are writing a method to select all objects with a status of Running, the name "FilterRunningObjects" more accurately describes the action of the method. This name implies that the method will return a collection of objects whose status is Running, effectively "filtering in" the objects of interest based on the specified criterion.

On the other hand, "FilterNonRunningObjects" suggests the opposite action: selecting objects that do not have the status of Running (i.e., objects that are either Idle or Completed). This would be misleading if the intended functionality is to select Running objects.

To summarize, you should go with "FilterRunningObjects" for clarity and adherence to common programming conventions regarding the term "filter." This naming convention makes it immediately clear to other developers what the method does: it filters the collection to include only those objects that are in the Running state."

Thanks again for the question, it prompted me to actually look into it.

File: src/modules/Elsa.Alterations/AlterationHandlers/ModifyVariableHandler.cs [hunk 3]
```diff
@@ -20,6 +19,7 @@ namespace Elsa.Alterations.AlterationHandlers;
 public class ModifyVariableHandler : AlterationHandlerBase<ModifyVariable>
 {
     /// <inheritdoc />
+    [RequiresUnreferencedCode("Calls System.Text.Json.JsonSerializer.Serialize<TValue>(TValue, JsonSerializerOptions)")]
```
[the remark this answers is not in the document]
Comment (sfmskywalker) [reply src/modules/Elsa.Alterations/AlterationHandlers/ModifyVariableHandler.cs:22]: Indeed. It's not what I want to do, but if we don't, the compiler generates a warning similar to the following:

image

This warning is part of the .NET platform's trimming feature, which tries to reduce the size of the compiled application by removing unused code. Please do let me know if there's a better way  to appease the compiler to not show this warning (besides ignoring this warning).

File: src/modules/Elsa.Alterations/AlterationHandlers/ModifyVariableHandler.cs [hunk 4]
```diff
@@ -32,39 +32,22 @@ protected override async ValueTask HandleAsync(AlterationContext context, Modify
...
-        var type = variableType.GenericTypeArguments[0];
-        return value.ConvertTo(type);
-    }
-
-    private async Task UpdateVariable(AlterationContext context, Variable variable, object? value, CancellationToken cancellationToken)
-    {
-        var variableStorage = variable.StorageDriverType ?? typeof(WorkflowStorageDriver);
-        var storageDriverManager = context.ServiceProvider.GetRequiredService<IStorageDriverManager>();
-        var storageDriver = storageDriverManager.Get(variableStorage)!;
         var activityExecutionContext = FindActivityExecutionContextContainingVariable(context, variable);
```
[the remark this answers is not in the document]
Comment (sfmskywalker) [reply src/modules/Elsa.Alterations/AlterationHandlers/ModifyVariableHandler.cs:42]: Sure, a variable that is visible to child scopes can be accessed by those scopes. But when it comes to updating a specific variable, only one scope can exist that defined the variable.

File: src/modules/Elsa.Workflows.Core/Activities/Flowchart/Activities/Flowchart.cs [hunk 5]
```diff
@@ -248,26 +254,31 @@ private async ValueTask OnChildCompletedAsync(ActivityCompletedContext context)
...
+    }
 
-                if (!hasPendingWork)
-                {
-                    logger.LogDebug("No pending work found");
-                    var hasFaultedActivities = flowchartContext.GetActiveChildren().Any(x => x.Status == ActivityStatus.Faulted);
+    private async Task CompleteIfNoPendingWorkAsync(ActivityExecutionContext context)
+    {
+        var logger = context.GetRequiredService<ILogger<Flowchart>>();
+        var flowchartContext = context;
```
[the remark this answers is not in the document]
Comment (sfmskywalker) [reply src/modules/Elsa.Workflows.Core/Activities/Flowchart/Activities/Flowchart.cs:267]: Fair question - I don't see any.

File: src/modules/Elsa.Workflows.Core/Services/WorkflowRunner.cs (hunk 1 above)
Comment (sfmskywalker) [reply src/modules/Elsa.Workflows.Core/Services/WorkflowRunner.cs:70]: Really - I didn't know. I will try.

Comment (sfmskywalker) [reply src/modules/Elsa.Workflows.Core/Services/WorkflowRunner.cs:70]: You are right, that worked! Let's remove the attribute whenever we come across it and fix it with the new() constraint.
